Lshire swift settings

AttilaTheFun opened the pull request.

cgrindel commented on examples/swift_protobuf_example/.bazelrc:
Please update this file to include:

```
# Import Shared settings
import %workspace%/../../shared.bazelrc

# Import CI settings.
import %workspace%/../../ci.bazelrc

# Try to import a local.rc file; typically, written by CI
try-import %workspace%/../../local.bazelrc
```

cgrindel commented on examples/swift_protobuf_example/BUILD.bazel:
Is this file necessary, if it is empty?

AttilaTheFun replied:
I can add a swift_library that depends on SwiftProtobuf and make the do_test script build it.

cgrindel commented on examples/swift_protobuf_example/MODULE.bazel:
```diff
@@ -0,0 +1,67 @@
+    path = "../..",
+)
+
+# swift_deps START
```
You can remove the `# swift_deps START/END` lines. They are no longer necessary.

cgrindel commented on examples/swift_protobuf_example/Package.swift:
```diff
@@ -0,0 +1,15 @@
+let package = Package(
+    name: "swift_protobuf_example",
+    dependencies: [
+        // These are the versions used by rules_swift
+        // .package(url: "https://github.com/grpc/grpc-swift-extras", exact: "1.0.0-beta.2"),
+        // .package(url: "https://github.com/grpc/grpc-swift.git", exact: "2.0.0-beta.2"), // Required by grpc-swift-extras @ 1.0.0-beta.2
+        // .package(url: "https://github.com/grpc/grpc-swift-protobuf.git", exact: "1.0.0-beta.2"), // Required by grpc-swift @ 2.0.0-beta.2
+        // .package(url: "https://github.com/grpc/grpc-swift-nio-transport.git", exact: "1.0.0-beta.2"), // Required by grpc-swift @ 2.0.0-beta.2
```
Do we need these comments? It mostly looks like commented code.

AttilaTheFun replied:
No I can remove them

cgrindel commented on swiftpkg/internal/swiftpkg_build_files.bzl:
Can we add [a unit test or two](https://github.com/cgrindel/rules_swift_package_manager/blob/ac2ba845c7760366bffbdffa120972ade731624a/swiftpkg/tests/swiftpkg_build_files_tests.bzl#L487) exercising this feature?

AttilaTheFun replied:
Ok I updated the test to verify that the SwiftSettings are translated to the following features:

```
    features = [
        "swift.enable_v6",
        "swift.experimental.BuiltinModule",
        "swift.upcoming.ExistentialAny",
    ],
```

cgrindel commented on examples/swift_protobuf_example/MODULE.bazel:
We need to add this new example to [example_infos.bzl](https://github.com/cgrindel/rules_swift_package_manager/blob/ac2ba845c7760366bffbdffa120972ade731624a/examples/example_infos.bzl#L129-L132). Once it is added there, you will need to run `//:tidy` so that the CI workflow is regenerated.

cgrindel reviewed (commented):
Overall, looks good. I left a few comments that we will need to address before we can merge.

AttilaTheFun commented:
@cgrindel Okay, I believe I've addressed all of your feedback.
CI is all green except for my new swift_protobuf_example which requires Swift 6.0 while CI is on an earlier swift version.
We could either merge it and upgrade CI later, or upgrade CI now so this will pass.

cgrindel commented:
@AttilaTheFun Based upon [this website](https://swiftversion.net/), it looks like we need Xcode 16.0 or higher for Swift 6. Do you agree?

AttilaTheFun commented:
@cgrindel yeah on macOS the swift version is tied to the Xcode version. 

Swift 6 released with Xcode 16, so we'll need to update the CI machines to build with Xcode 16.

On Ubuntu you can just install the desired swift version directly.

luispadron commented on swiftpkg/tests/swiftpkg_build_files_tests.bzl:
```diff
@@ -566,6 +570,11 @@ swift_binary(
         "@rules_swift_package_manager//config_settings/spm/configuration:release": ["-cross-module-optimization"],
         "//conditions:default": [],
     }),
+    features = [
+        "swift.enable_v6",
+        "swift.experimental.BuiltinModule",
+        "swift.upcoming.ExistentialAny",
+    ],
```
Should this be done via features or via `copts`? I think features is more readable but requires rules_swift to be updated for new ones where-as `copts` could support things we do not know of yet.

Thoughts @brentleyjones

luispadron replied:
> rules_swift won't need to update for new ones (the way it transforms it is dynamic). Given that, I think it doesn't matter? Features could be better if the copts ever change for how one would enable them, but I doubt that matters.

cgrindel reviewed (approved):
LGTM. I will try to refactor the CI workflow generation, this weekend.

cgrindel commented on examples/grpc_example/do_test:
```diff
@@ -7,5 +7,5 @@ set -o errexit -o nounset -o pipefail
 
 # Ensure that it builds and tests pass
-"${bazel}" test //...
+"${bazel}" build //...
```
Why did we switch away from test? We should at least try to run the thing that we just built.

AttilaTheFun replied:
The GRPC authors removed the TestClient mode between 1.x and 2.x
Now the only option is just starting a server and communicating on localhost.
I did in different terminals run the client and server and verify the message round tripped, but I'm not sure how to integrate that with bazel testing.

cgrindel commented on examples/grpc_example/sources/common/ClientRequest+Convenience.swift:
```diff
@@ -0,0 +1,12 @@
+// This file is copied from ClientRequest+Convenience.swift in GRPCCore
+// TODO: Request that they make this public.
```
Please do not leave TODO comments. If it is something that needs to be followed up, please create an issue and refer to it (e.g., GHxxx where xxx is the `rules_swift_package_manager`  issue number). If we are requesting something in another repo, which is the case here, please create an issue there and post a link in a comment here.

AttilaTheFun replied:
Sure I can, though this is just an example and not part of the core repo?
For whatever reason, it's currently difficult to make unary requests out-of-the-box in grpc-swift 2.0
Everything actually goes through bidirectional streaming and unary requests are just bidirectional streaming requests where the client makes one request and hangs up after the response. I can file an issue on their repo and link it here though.

AttilaTheFun replied:
Updated the comment to:
// Requested they make this public: https://github.com/grpc/grpc-swift/issues/2213
// When they do, we can remove this.

cgrindel commented on examples/grpc_example/sources/common/ClientResponse+Convenience.swift:
```diff
@@ -0,0 +1,105 @@
+// This file is copied from ClientResponse+Convenience.swift in GRPCCore
+// TODO: Request that they make this public.
```
TODO comment.

cgrindel reviewed (changes requested):
A couple of items before we can merge.

cgrindel commented:
@mergifyio queue
